src/site/content/en/blog/subsonic/index.md

```diff
@@ -0,0 +1,171 @@
+
+The easiest fix for this problem (which also gives us a few other nice advantages) is to package all the resource files into a single bundle, which we'll download with a single XHR call, which gives us the progress events we need to display a nice progress bar.
+
+Building a custom bundle file format is not terribly hard, and would even solve a few problems, but would require creating a tool for creating the bundle format.  An alternative solution is to use an existing archive format for which tools already exist, and then only need to write a decoder to run in the browser.  We don't need a compressed archive format because HTTP can already compress data using gzip or deflate algorithms just fine.  For these reasons, we settled on the TAR file format.
+
+TAR is a relatively simple format.  Every record (file) has a 512 byte header, followed by the file contents padded out to 512 bytes.  The header has only a few relevant or interesting fields for our purposes, mainly the file type and name, which are stored at fixed positions within the header.
+
+Header fields in the TAR format are stored at fixed locations with fixed sizes in the header block.  For example, the last modification timestamp of the file is stored at 136 bytes from the start of the header, and is 12 bytes long.  All numeric fields are encoded as octal numbers stored in ASCII format.  To parse the fields, then, we just extract the fields from our array buffer, and for numeric fields we call parseInt() being sure to pass in the second parameter to indicate the desired octal base.
+
+One of the most important fields is the type field.  This is a single digit octal number which tells us what type of file the record contains.  The only two interesting record types for our purposes are regular files ('0') and directories ('5').  If we were dealing with arbitrary TAR files we might also care about symbolic links ('2') and possibly hard links ('1').
+
+Each header is followed immediately by the contents of the file described by the header (excepting file types that have no contents of their own, like directories).  The file contents are then followed by padding to ensure that every header begins on a 512-byte boundary.  Thus, to calculate the total length of a file record in a TAR file, we first must read the header for the the file.  We then add the length of the header (512 bytes) with the length of the file contents extracted from the header.  Finally, we add any padding bytes necessary to make the offset align to 512 bytes, which can be done easily by dividing the file length by 512, taking the ceiling of the number, and then multiplying by 512.
```

Comment:
```suggestion
One problem with chained loading of resources is that there's no way to get a complete byte count of all the data. The consequence of this is that there's no way to make a simple, reliable progress dialog for downloads. Since we're going to be downloading all the content and caching it, and this can take a rather long time for larger games, giving the player a nice progress dialog is pretty important.

The easiest fix for this problem (which also gives us a few other nice advantages) is to package all the resource files into a single bundle, which we'll download with a single XHR call, which gives us the progress events we need to display a nice progress bar.

Building a custom bundle file format is not terribly hard, and would even solve a few problems, but would require creating a tool for creating the bundle format. An alternative solution is to use an existing archive format for which tools already exist, and then need to write a decoder to run in the browser. We don't need a compressed archive format because HTTP can already compress data using gzip or deflate algorithms just fine. For these reasons, we settled on the TAR file format.

TAR is a relatively simple format. Every record (file) has a 512 byte header, followed by the file contents padded out to 512 bytes. The header has only a few relevant or interesting fields for our purposes, mainly the file type and name, which are stored at fixed positions within the header.

Header fields in the TAR format are stored at fixed locations with fixed sizes in the header block. For example, the last modification timestamp of the file is stored at 136 bytes from the start of the header, and is 12 bytes long. All numeric fields are encoded as octal numbers stored in ASCII format. To parse the fields, then, we extract the fields from our array buffer, and for numeric fields we call `parseInt()` being sure to pass in the second parameter to indicate the desired octal base.

One of the most important fields is the type field. This is a single digit octal number which tells us what type of file the record contains. The only two interesting record types for our purposes are regular files (`'0'`) and directories (`'5'`). If we were dealing with arbitrary TAR files we might also care about symbolic links (`'2'`) and possibly hard links (`'1'`).

Each header is followed immediately by the contents of the file described by the header (excepting file types that have no contents of their own, like directories). The file contents are then followed by padding to ensure that every header begins on a 512-byte boundary. Thus, to calculate the total length of a file record in a TAR file, we first must read the header for the the file. We then add the length of the header (512 bytes) with the length of the file contents extracted from the header. Finally, we add any padding bytes necessary to make the offset align to 512 bytes, which can be done easily by dividing the file length by 512, taking the ceiling of the number, and then multiplying by 512.
```